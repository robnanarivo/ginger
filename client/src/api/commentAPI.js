import axios from 'axios';
import domain from '../constants/profile';

const createComment = async (comment) => {
  const token = sessionStorage.getItem('userToken');
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const res = await axios.post(`${domain}/api/comments/`, comment, { headers: config });
    return res.data;
  } catch (err) {
    return err.response;
  }
};

const deleteComment = async (commentId) => {
  const token = sessionStorage.getItem('userToken');
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const res = await axios.delete(`${domain}/api/comments/${commentId}`, { headers: config });
    return res.data;
  } catch (err) {
    return err.response;
  }
};

const updateComment = async (commentId, comment) => {
  const token = sessionStorage.getItem('userToken');
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const res = await axios.put(`${domain}/api/comments/${commentId}`, comment, { headers: config });
    return res.data;
  } catch (err) {
    if (err.response) {
      return err.response.data;
    }
    return { error: 'Cannot connect to server' };
  }
};

const getPost = async (postId) => {
  try {
    const res = await axios.get(`${domain}/api/posts/${postId}`);
    const { data } = res;
    const { content } = data;
    if (content.audio) {
      Object.assign(content, { audio: `${domain}/api/multimedia/${content.audio}` });
    }
    if (content.video) {
      Object.assign(content, { video: `${domain}/api/multimedia/${content.video}` });
    }
    if (content.image) {
      Object.assign(content, { image: `${domain}/api/multimedia/${content.image}` });
    }
    return data;
  } catch (err) {
    return err.response;
  }
};

export {
  createComment,
  deleteComment,
  updateComment,
  getPost,
};
