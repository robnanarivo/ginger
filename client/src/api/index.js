import {
  getListOfGroups,
  getGroupData,
  updateGroup,
  removeAdmin,
  getGroupAdmins,
  createGroup,
} from './fetcher';
import login from './login';
import register from './register';
import { getNotifications, updateNotification } from './notificationAPI';
import getUsers from './user';
import { updateProfile, getProfile } from './profileAPI';
import {
  createComment, deleteComment, updateComment, getPost,
} from './commentAPI';

export {
  getGroupData,
  getListOfGroups,
  updateGroup,
  getGroupAdmins,
  removeAdmin,
  login,
  register,
  createGroup,
  getNotifications,
  updateNotification,
  getUsers,
  updateProfile,
  getProfile,
  createComment,
  deleteComment,
  updateComment,
  getPost,
};
