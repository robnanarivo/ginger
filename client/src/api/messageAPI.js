import axios from 'axios';
import domain from '../constants/profile';

const sendMessage = async (message) => {
  try {
    const res = await axios.post(`${domain}/api/messages`, message);
    return res.data;
  } catch (err) {
    return err.response;
  }
};

const getMessageHistory = async (chatter1, chatter2, offset) => {
  const params = {
    chatter1,
    chatter2,
    offset,
    limit: 10,
  };
  try {
    const res = await axios.get(`${domain}/api/messages`, { params });
    return res.data;
  } catch (err) {
    return err.response;
  }
};

const getNewMessage = async (chatter1, chatter2) => {
  const params = {
    chatter1,
    chatter2,
    offset: 0,
    limit: 1,
  };
  try {
    const res = await axios.get(`${domain}/api/messages`, { params });
    return res.data;
  } catch (err) {
    return err.response;
  }
};

export {
  sendMessage,
  getMessageHistory,
  getNewMessage,
};
