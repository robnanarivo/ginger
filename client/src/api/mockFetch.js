export const groupData = {
  topics: ['Sports', 'Universe', 'Science', 'Astronomy'],
  groupInfo: { createOn: 'Feb 20 2008', member: '2.8k' },
  groupId: '11',
  admins: [],
  users: [],
  posts: [
    {
      postId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      groupId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      userId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      content: {
        title: 'Stars are so beautiful',
        text: 'star!',
        image: '/starPic.jpeg',
        audio: '',
        video: '',
      },
      date: 'September 16, 2021',
      commentCount: 30,
      likeCount: 8750,
      avatar: 'AB',
      timeStamp: {},
      comments: [
        {
          commentId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          userId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          content: 'somewhat random text',
          timeStamp: {},
          parentPostId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
        },
      ],
      flagedBy: [
        'd290f1ee-6c54-4b01-90e6-d701748f0851',
      ],
      hiddenBy: [
        'd290f1ee-6c54-4b01-90e6-d701748f0851',
      ],
    },
    {
      postId: 'd290f1ee-6c54-4b01-90e6-d701748f0852',
      groupId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      userId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      content: {
        title: 'Look at the Galaxy!',
        text: 'beautiful night',
        image: '/galaxyPic.jpeg',
        audio: '',
        video: '',
      },
      timeStamp: {},
      date: 'September 17, 2021',
      commentCount: 328,
      likeCount: 18650,
      avatar: 'NK',
      comments: [
        {
          commentId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          userId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          content: 'somewhat random text',
          timeStamp: {},
          parentPostId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
        },
      ],
      flagedBy: [
        'd290f1ee-6c54-4b01-90e6-d701748f0851',
      ],
      hiddenBy: [
        'd290f1ee-6c54-4b01-90e6-d701748f0851',
      ],
    },
  ],
  groupName: 'Galaxy',
  userStatus: 'joined',
};

export const listOfAdmins = [
  {
    userId: '11',
    userName: 'Alex',
    profilePicture: '',
  },
  {
    userId: '12',
    userName: 'Tom',
    profilePicture: '',
  },
];

export const listOfGroupAll = [
  {
    groupId: '11',
    groupName: 'Galaxy',
    groupPic: '',
    member: '40',
    userStatus: 'joined',
  },
  {
    groupId: '2',
    groupName: 'SEPTA',
    groupPic: '',
    member: '40',
    userStatus: 'not joined',
  },
  {
    groupId: '3',
    groupName: 'Study Group',
    groupPic: '',
    member: '48920',
    userStatus: 'joined',
  },
  {
    groupId: '4',
    groupName: 'Involution',
    groupPic: '',
    member: '345781',
    userStatus: 'pending',
  },
  {
    groupId: '5',
    groupName: 'Ubuntu',
    groupPic: '',
    member: '908907',
    userStatus: 'pending',
  },
  {
    groupId: '6',
    groupName: 'Debian',
    groupPic: '',
    member: '462',
    userStatus: 'not joined',
  },
  {
    groupId: '7',
    groupName: 'CentOS',
    groupPic: '',
    member: '471',
    userStatus: 'joined',
  },
  {
    groupId: '8',
    groupName: 'Assembly',
    groupPic: '',
    member: '589',
    userStatus: 'not joined',
  },
  {
    groupId: '9',
    groupName: 'JIT',
    groupPic: '',
    member: '25',
    userStatus: 'not joined',
  },
  {
    groupId: '10',
    groupName: 'NLP',
    groupPic: '',
    member: '5',
    userStatus: 'not joined',
  },
];

export const listOfGroupPub = [
  {
    groupId: '11',
    groupName: 'Galaxy',
    groupPic: '',
    groupInfo: { createOn: '', member: '40' },
    userStatus: 'joined',
  },
  {
    groupId: '2',
    groupName: 'SEPTA',
    groupPic: '',
    groupInfo: { createOn: '', member: '3640' },
    userStatus: 'not join',
  },
  {
    groupId: '3',
    groupName: 'Study Group',
    groupPic: '',
    groupInfo: { createOn: '', member: '48920' },
    userStatus: 'joined',
  },
  {
    groupId: '4',
    groupName: 'Involution',
    groupPic: '',
    groupInfo: { createOn: '', member: '345781' },
    userStatus: 'pending',
  },
  {
    groupId: '5',
    groupName: 'Ubuntu',
    groupPic: '',
    groupInfo: { createOn: '', member: '908907' },
    userStatus: 'pending',
  },
  {
    groupId: '6',
    groupName: 'Debian',
    groupPic: '',
    groupInfo: { createOn: '', member: '462' },
    userStatus: 'notJoined',
  },
];
