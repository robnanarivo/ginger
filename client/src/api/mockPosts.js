const postData = {
  topics: ['Sports', 'Universe', 'Science', 'Astronomy'],
  groupInfo: { createOn: 'Feb 20 2008', member: '2.8k' },
  groupId: '11',
  posts: [
    {
      postId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      groupId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
      creatorId: '61b26b150c01529bed74c6f0',
      creatorName: 'Wang',
      title: 'Stars are so beautiful',
      content: {
        text: 'star!',
        image: '/starPic.jpeg',
        audio: '',
        video: '',
      },
      profilePicture: '/starPic.jpeg',
      timestamp: 1639429592291,
      comments: [
        {
          commentId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          userId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
          content: 'somewhat random text',
          timeStamp: 1639429592456,
          parentPostId: 'd290f1ee-6c54-4b01-90e6-d701748f0851',
        },
      ],
      flagedBy: [
        'd290f1ee-6c54-4b01-90e6-d701748f0851',
      ],
      hiddenBy: [
        'd290f1ee-6c54-4b01-90e6-d701748f0851',
      ],
    },
  ],
  groupName: 'Galaxy',
  userStatus: 'joined',
};

export default postData;
