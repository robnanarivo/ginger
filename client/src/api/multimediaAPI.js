import axios from 'axios';
import domain from '../constants/profile';

const upload = async (file) => {
  try {
    const formData = new FormData();
    formData.append('file', file);
    const res = await axios.post(`${domain}/api/multimedia`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return res.data;
  } catch (err) {
    return err.response;
  }
};

export default upload;
