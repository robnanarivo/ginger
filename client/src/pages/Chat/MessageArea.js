import { React, useState, useEffect } from 'react';
import { Typography, Grid } from '@mui/material';
import InfiniteScroll from 'react-infinite-scroll-component';
import ChatMessage from './ChatMessage';
import ChatImage from './ChatImage';
import ChatAudio from './ChatAudio';
import ChatVideo from './ChatVideo';
import { getMessageHistory, getNewMessage } from '../../api/messageAPI';
import domain from '../../constants/profile';

const serverURL = `${domain}/api/multimedia/`;

const MessageArea = ({ chatter, chatterId }) => {
  const myUserId = sessionStorage.getItem('userId');

  const [messages, setMessages] = useState([]);
  const [prevTimestamp, setPrevTimestamp] = useState();
  const [offset, setOffset] = useState(10);
  const [hasMore, setHasMore] = useState(true);
  const [finishSetUp, setFinishSetUp] = useState(false);

  const fetchData = async () => {
    const messageHistory = await getMessageHistory(myUserId, chatterId, offset);
    if (messageHistory.length === 0) {
      setHasMore(false);
    }
    setOffset(offset + 10);
    setMessages(messages.concat(messageHistory));
  };

  // initial set-up
  useEffect(async () => {
    const messageHistory = await getMessageHistory(myUserId, chatterId, 0);
    if (messageHistory.length === 0) {
      setHasMore(false);
      setPrevTimestamp(0);
    } else {
      setMessages(messageHistory);
      setPrevTimestamp(messageHistory[0].timestamp);
    }
    setFinishSetUp(true);
  }, []);

  const updateMessage = async () => {
    const newMessage = await getNewMessage(myUserId, chatterId);
    if (prevTimestamp !== undefined
      && newMessage.length > 0
      && newMessage[0].timestamp > prevTimestamp) {
      setPrevTimestamp(newMessage[0].timestamp);
    }
  };

  useEffect(() => {
    let newPolling;
    if (finishSetUp) {
      newPolling = setInterval(updateMessage, 3000);
    }
    return () => clearInterval(newPolling);
  }, [finishSetUp]);

  useEffect(async () => {
    if (finishSetUp) {
      const newMessage = await getNewMessage(myUserId, chatterId);
      setMessages((messageHistory) => newMessage.concat(messageHistory));
      setOffset(offset + 1);
    }
  }, [prevTimestamp]);

  const Loader = () => (
    <Grid container alignItems="center" justifyContent="center">
      <Typography variant="h5" sx={{ color: 'gray' }}>Loading...</Typography>
    </Grid>
  );

  const messageRenderer = (message) => {
    const time = new Date(message.timestamp);
    const timeString = time.toLocaleString();
    let toRender;
    if (message.message.contentType === 'text') {
      toRender = (
        <ChatMessage
          key={message._id}
          sender={message.sender}
          content={message.message.content}
          timestamp={timeString}
          isSelf={message.sender === myUserId}
        />
      );
    } else if (message.message.contentType === 'image') {
      toRender = <ChatImage key={message._id} chatter={chatter} url={`${serverURL}${message.message.content}`} timestamp={timeString} isSelf={message.sender === myUserId} />;
    } else if (message.message.contentType === 'audio') {
      toRender = <ChatAudio key={message._id} chatter={chatter} url={`${serverURL}${message.message.content}`} timestamp={timeString} isSelf={message.sender === myUserId} />;
    } else if (message.message.contentType === 'video') {
      toRender = <ChatVideo key={message._id} chatter={chatter} url={`${serverURL}${message.message.content}`} timestamp={timeString} isSelf={message.sender === myUserId} />;
    }
    return (toRender);
  };

  return (
    <div>
      <div
        id="scrollableDiv"
        style={{
          height: 370, overflow: 'auto', display: 'flex', flexDirection: 'column-reverse',
        }}
      >
        <InfiniteScroll
          dataLength={messages.length}
          style={{ display: 'flex', flexDirection: 'column-reverse' }}
          next={fetchData}
          inverse
          hasMore={hasMore}
          loader={<Loader />}
          scrollableTarget="scrollableDiv"
        >
          {messages.map(messageRenderer)}
        </InfiniteScroll>
      </div>
    </div>
  );
};

export default MessageArea;
