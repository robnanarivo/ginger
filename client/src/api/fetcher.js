// import { groupData, listOfGroupAll, listOfGroupPub } from './mockFetch';
import axios from 'axios';
import { listOfAdmins } from './mockFetch';

const domain = !process.env.NODE_ENV || process.env.NODE_ENV === 'development'
  ? 'http://localhost:5000'
  : '';

const getGroupData = async (groupId) => {
  const token = sessionStorage.getItem('userToken');
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const ret = await axios.get(`${domain}/api/group/${groupId}`, { headers: config });
    const { data } = ret;
    if (data.groupIcon !== '') {
      Object.assign(data, { groupIcon: `${domain}/api/multimedia/${data.groupIcon}` });
    } else {
      Object.assign(data, { groupIcon: '' });
    }
    // data.posts.forEach((post) => {
    //   const { content } = post;
    //   if (content.audio) {
    //     Object.assign(content, { audio: `${domain}/api/multimedia/${content.audio}` });
    //   }
    //   if (content.video) {
    //     Object.assign(content, { video: `${domain}/api/multimedia/${content.video}` });
    //   }
    //   if (content.image) {
    //     Object.assign(content, { image: `${domain}/api/multimedia/${content.image}` });
    //   }
    // });
    return data;
  } catch (err) {
    return err.response;
  }
};

const createGroup = async (groupInfo) => {
  const token = sessionStorage.getItem('userToken');
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const res = await axios.post(`${domain}/api/group/`, groupInfo, { headers: config });
    return res.data;
  } catch (err) {
    // console.log(err);
    return err.response.data;
  }
};

const updateGroup = async (groupId, action, userIds, adminIds) => {
  const token = sessionStorage.getItem('userToken');
  const updateInfo = { action, userIds, adminIds };
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const res = await axios.put(`${domain}/api/group/${groupId}`, updateInfo, { headers: config });
    // console.log(res);
    if (res.data.results === 'success') {
      return true;
    }
    return false;
  } catch (err) {
    // console.log(err);
    return false;
  }
};

const getListOfGroups = async (topic) => {
  const token = sessionStorage.getItem('userToken');
  try {
    const config = {
      authorization: `Bearer ${token}`,
    };
    const ret = await axios.get(`${domain}/api/groups/${topic}`, { headers: config });
    const { data } = ret;
    data.forEach((d) => {
      if (d.groupIcon !== '') {
        Object.assign(d, { groupIcon: `${domain}/api/multimedia/${d.groupIcon}` });
      } else {
        Object.assign(d, { groupIcon: '' });
      }
    });
    return ret.data;
  } catch (err) {
    // console.log(err.response);
    return err.response;
  }
};

// const getGroupAdmins = async (groupId) => {
//   const groupInfo = ['Alex', 'Tom'];
//   if (groupId) {
//     return listOfAdmins;
//   }
//   // const token = sessionStorage.getItem('userToken');
//   // try {
//   //   const config = {
//   //     authorization: `Bearer ${token}`,
//   //   };
//   //   const ret = await axios.get(`${domain}/api/group/${groupId}`, { headers: config });
//   //   const { data } = ret;
//   //   console.log(data);
//   // } catch (err) {
//   //   return err.response;
//   // }
//   return groupInfo;
// };

const getGroupAdmins = async (groupId) => {
  const groupInfo = ['Alex', 'Tom'];
  if (groupId) {
    return listOfAdmins;
  }
  // const token = sessionStorage.getItem('userToken');
  // try {
  //   const config = {
  //     authorization: `Bearer ${token}`,
  //   };
  //   const ret = await axios.get(`${domain}/api/group/${groupId}`, { headers: config });
  //   const { data } = ret;
  //   console.log(data);
  // } catch (err) {
  //   return err.response;
  // }
  return groupInfo;
};

const removeAdmin = async (adminId) => {
  if (!adminId) {
    return true;
  }
  return true;
};

export {
  getGroupData,
  getListOfGroups,
  createGroup,
  updateGroup,
  getGroupAdmins,
  removeAdmin,
};
