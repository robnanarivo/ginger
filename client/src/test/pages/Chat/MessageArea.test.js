import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import MessageArea from '../../../pages/Chat/MessageArea';
import { getMessageHistory } from '../../../api/messageAPI';
import 'jest-styled-components';

jest.mock('../../../api/messageAPI');

describe('MessageArea unit test', () => {
  beforeEach(() => {
    const messages = [{
      _id: 123,
      timestamp: 0,
      sender: 'bob',
      message: {
        content: 'I am good',
        contentType: 'text',
      },
    }];
    getMessageHistory.mockResolvedValue(messages);
  });

  it('Mock new message', async () => {
    await act(async () => {
      render(<MessageArea />);
    });
    expect(screen.queryByText(/good/i)).toBeInTheDocument();
  });

  it('MessageArea Snapshot test', () => {
    const tree = renderer
      .create(<MessageArea />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
