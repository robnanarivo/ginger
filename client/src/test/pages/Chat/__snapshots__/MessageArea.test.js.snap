// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageArea unit test MessageArea Snapshot test 1`] = `
<div>
  <div
    id="scrollableDiv"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column-reverse",
        "height": 370,
        "overflow": "auto",
      }
    }
  >
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component "
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "display": "flex",
            "flexDirection": "column-reverse",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="MuiGrid-root MuiGrid-container css-go2vvz-MuiGrid-root"
        >
          <h5
            className="MuiTypography-root MuiTypography-h5 css-o50h7g-MuiTypography-root"
          >
            Loading...
          </h5>
        </div>
      </div>
    </div>
  </div>
</div>
`;
